README.md
# scrollable_panel
[![pub package](https://img.shields.io/pub/v/scrollable_panel.svg)](https://pub.dev/packages/scrollable_panel)

drag to expand and then can scroll contents.
similar "Nearby spots" panel on google map app.

https://pub.dev/packages/scrollable_panel

![](https://github.com/renoinn/scrollable_panel/blob/master/panel_movie.gif)

## Panel Properties

| Properties | Data Type | Description |
|--|--|--|
| builder | ScrollableWidgetBuilder | builder for inner panel view. ScrollableWidgetBuilder type define in draggable_scrollable_sheet.dart. provide PanelScrollController that extends ScrollController. |
| controller | PanelController | control panel from some interaction like button. |
| defaultPanelState | PanelState | set default panel state. (default value PanelState.open) |
| defaultPanelSize | double | set default panel height factor (default value 0.25) |
| minPanelSize | double | set minimum panel height factor. (default value 0) |
| maxPanelSize | double | set maximum panel height factor. (default value 1.0) |
| onOpen | VoidCallback | if non-null, this callback is called when panel state become open |
| onClose | VoidCallback | if non-null, this callback is called when panel state become close |
| onExpand | VoidCallback | if non-null, this callback is called when panel state become expand |

## PanelController actions

| Action | Data Type | Description |
|--|--|--|
| open | void | open panel. panel animate to defaultPanelSize. |
| expand | void | expand panel. panel animate to maxPanelSize. |
| close | void | close panel. panel animate to minPanelSize. |
| animateTo | double | animate panel to value height. |

## Usage

```dart
class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  PanelController _panelController = PanelController();
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('scrollable panel'),
      ),
      body: Stack(
        children: <Widget>[
          InkWell(
            onTap: () {
              _panelController.toDefault();
            },
            child: _FirstView(),
          ),
          ScrollablePanel(
            controller: _panelController,
            builder: (context, controller) {
              return SingleChildScrollView(
                controller: controller,
                child: _SecondView(),
              );
            },
          ),
        ],
      ),
    );
  }
}
```

## TODO

- ~write README.md and dartdoc~
- ~null-safety support~
- ~write test code~
- add `List<double> anchor` property to use multi snapping point.


CHANGELOG.md
## [0.4.1] - 2021/03/12

* add getter isOpen, isClose, isExpand
* add test code

## [0.4.0] - 2021/03/04

* null safety support
required dart version `>=2.12.0`

## [0.3.0] - 2021/03/03

* add parameter onOpen, defaultPanelState
* add open() method
* add dartdoc

## [0.2.2] - 2020/09/11

* add onClose, onExpand
* WillPopScope

## [0.2.1] - 2020/09/11

* add animation getter

## [0.2.0] - 2020/09/10

* update panel usage

## [0.1.2] - 2019/10/29

* fix negative velocity

## [0.1.1] - 2019/10/23

* delete debug log

## [0.1.0] - 2019/10/22

* add velocity control and defaultPanelSize

## [0.0.3] - 2019/10/18

* fix bug and example

## [0.0.2] - 2019/10/18

* fix bug and example

## [0.0.1] - 2019/10/09

* initial release.


test/scrollable_panel_test.dart
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:scrollable_panel/scrollable_panel.dart';

void main() {
  group('PanelController test', () {
    test('ScrollablePanelState not attached', () {
      final panelController = PanelController();
      expect(panelController.isAttached, false);
    });

    testWidgets('ScrollablePanelState is attached', (tester) async {
      final scrollablePanelKey = GlobalKey<ScrollablePanelState>();
      final panelController = PanelController();
      await tester.pumpWidget(ScrollablePanel(
        key: scrollablePanelKey,
        controller: panelController,
        builder: (context, controller) => SingleChildScrollView(
          controller: controller,
        ),
      ));

      expect(panelController.isAttached, true);
      expect(panelController.isOpen, true);
    });

    testWidgets('PanelState is Expand call after panelController.expand()', (tester) async {
      final scrollablePanelKey = GlobalKey<ScrollablePanelState>();
      final panelController = PanelController();
      await tester.pumpWidget(ScrollablePanel(
        key: scrollablePanelKey,
        controller: panelController,
        builder: (context, controller) => SingleChildScrollView(
          controller: controller,
          child: Directionality(
            textDirection: TextDirection.ltr,
            child: Column(
              children: [
                for (int i = 0; i < 20; i++) Text(i.toString()),
              ],
            ),
          ),
        ),
      ));

      panelController.expand();
      await tester.pumpAndSettle(const Duration(milliseconds: 300));
      expect(panelController.isExpand, true);
    });
  });
}


example/scrollable_panel_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:scrollable_panel_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/panel.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// TODO List<double> anchorみたいなのを用意して、そのポイントでsnapするようにしたい。open()した時に、defaultPanelSizeが指定されていればdefaultPanelSizeに、されていなければanchor[0]になる。toNext/toPrevで上下する。

class ScrollablePanel extends StatefulWidget {
  const ScrollablePanel({
    Key? key,
    required this.builder,
    this.controller,
    this.defaultPanelState = PanelState.open,
    this.defaultPanelSize = 0.25,
    this.minPanelSize = 0,
    this.maxPanelSize = 1.0,
    this.onOpen,
    this.onClose,
    this.onExpand,
  })  : assert(minPanelSize <= defaultPanelSize),
        assert(defaultPanelSize <= maxPanelSize),
        assert(minPanelSize < maxPanelSize),
        super(key: key);

  /// default PanelState. if null, set [PanelState.open].
  final PanelState defaultPanelState;

  /// default panel height. when open panel, animate to defaultPanelSize.
  /// set between 0.0 to 1.0 .
  final double defaultPanelSize;

  /// minimum panel height. when close panel, animate to minPanelSize.
  /// set between 0.0 to 1.0 .
  final double minPanelSize;

  /// maximum panel height. when expand panel, animate to defaultPanelSize.
  /// set between 0.0 to 1.0 .
  final double maxPanelSize;

  /// build inner widget in panel.
  /// pass [PanelScrollController] instance to scrollController.
  ///
  /// ```dart
  /// builder: (context, controller) {
  ///   return SingleChildScrollView(
  ///     controller: controller,
  ///     child: InnerView(),
  ///   );
  /// },
  ///
  /// ```
  ///
  /// can't drag panel if builder callback return non [Scrollable] widget.
  /// if you want drag to expand, need to return [Scrollable] widget.
  final ScrollableWidgetBuilder builder;

  /// [PanelController] instance for control panel.
  final PanelController? controller;

  /// call this method when PanelState become open if not null.
  final VoidCallback? onOpen;

  /// call this method when PanelState become close if not null.
  final VoidCallback? onClose;

  /// call this method when PanelState become expand if not null.
  final VoidCallback? onExpand;

  @override
  State<StatefulWidget> createState() => ScrollablePanelState();
}

enum PanelState {
  /// [PanelState.open] when panel height value greater-than 0.0 and less-than 1.0
  open,

  /// [PanelState.expand] when panel height value equal maxPanelSize.
  expand,

  /// [PanelState.close] when panel height value equal minPanelSize.
  close,
}

class ScrollablePanelState extends State<ScrollablePanel> with SingleTickerProviderStateMixin {
  final double _snapThreshold = 0.2;
  double get defaultPanelSize => widget.defaultPanelSize;
  double get minPanelSize => widget.minPanelSize;
  double get maxPanelSize => widget.maxPanelSize;
  late ScrollController _scrollController;
  late AnimationController _dragAnimationController;
  late PanelState _panelState;

  @override
  void initState() {
    super.initState();
    _panelState = widget.defaultPanelState;
    var startValue = defaultPanelSize;
    if (widget.defaultPanelState == PanelState.close) {
      startValue = minPanelSize;
    } else if (widget.defaultPanelState == PanelState.expand) {
      startValue = maxPanelSize;
    }
    _dragAnimationController = AnimationController(value: startValue, vsync: this, duration: const Duration(milliseconds: 300));
    _dragAnimationController.addListener(() {
      if (mounted) {
        setState(() {});

        final onOpen = widget.onOpen;
        if (_dragAnimationController.value == defaultPanelSize) {
          if (onOpen != null && _panelState != PanelState.open) {
            onOpen();
          }
          _panelState = PanelState.open;
        }

        final onClose = widget.onClose;
        if (_dragAnimationController.value == minPanelSize) {
          if (onClose != null && _panelState != PanelState.close) {
            onClose();
          }
          _panelState = PanelState.close;
        }

        final onExpand = widget.onExpand;
        if (_dragAnimationController.value == maxPanelSize) {
          if (onExpand != null && _panelState != PanelState.expand) {
            onExpand();
          }
          _panelState = PanelState.expand;
        }
      }
    });
    final panelController = widget.controller ?? PanelController();
    panelController.addState(this);
    _scrollController = PanelScrollController(controller: panelController);
  }

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: Alignment.bottomCenter,
      child: WillPopScope(
        onWillPop: _onWillPop,
        child: FractionallySizedBox(
          heightFactor: _dragAnimationController.value,
          alignment: Alignment.bottomCenter,
          child: NotificationListener<ScrollNotification>(
            onNotification: _onScroll,
            child: widget.builder(context, _scrollController),
          ),
        ),
      ),
    );
  }

  void _onDragEnd() {
    final fromMinValue = minPanelSize - _dragAnimationController.value;
    if (fromMinValue.abs() < (defaultPanelSize - minPanelSize) / 2) {
      _animateTo(minPanelSize);
      return;
    }

    final half = (maxPanelSize - defaultPanelSize) / 2 + defaultPanelSize;
    final fromMaxValue = maxPanelSize - _dragAnimationController.value;
    if (fromMaxValue.abs() < _snapThreshold || _dragAnimationController.value > half) {
      _animateTo(maxPanelSize);
      return;
    }

    final fromDefaultValue = defaultPanelSize - _dragAnimationController.value;
    if (fromDefaultValue.abs() < _snapThreshold || _dragAnimationController.value < half) {
      _animateTo(defaultPanelSize);
      return;
    }
  }

  bool _onScroll(ScrollNotification notification) {
    if (notification is ScrollEndNotification) {
      _onDragEnd();
    }

    return true;
  }

  Future<void> _animateTo(double to) async {
    await _dragAnimationController.animateTo(to);
  }

  Future<void> _toDefault() async {
    await _animateTo(defaultPanelSize);
  }

  Future<void> _expand() async {
    await _animateTo(maxPanelSize);
  }

  Future<void> _close() async {
    await _animateTo(minPanelSize);
  }

  Future<bool> _onWillPop() async {
    _scrollController.jumpTo(0.0);
    if (_dragAnimationController.value > defaultPanelSize) {
      await _toDefault();
      return false;
    } else if (_dragAnimationController.value > minPanelSize) {
      await _close();
      return false;
    }
    return true;
  }
}

class PanelController {
  PanelController();

  ScrollablePanelState? _state;

  /// ValueListenable panel animation. when panel drag or call open/close/expand, notifier listeners.
  Animation? get animation => _state?._dragAnimationController;

  /// PanelController need attach _ScrollablePanelState.
  bool get isAttached => _state != null;

  /// return defaultPanelSize
  double get defaultPanelSize => _state?.defaultPanelSize ?? 0.25;

  /// return minPanelSize
  double get minPanelSize => _state?.minPanelSize ?? 0;

  /// return maxPanelSize
  double get maxPanelSize => _state?.maxPanelSize ?? 1.0;

  /// return true if panel state is [PanelState.open]
  bool get isOpen => _state?._panelState == PanelState.open;

  /// return true if panel state is [PanelState.close]
  bool get isClose => _state?._panelState == PanelState.close;

  /// return true if panel state is [PanelState.expand]
  bool get isExpand => _state?._panelState == PanelState.expand;

  @visibleForTesting
  void addState(ScrollablePanelState state) {
    _state = state;
  }

  void _updateExtent(double delta) {
    if (!isAttached) return;
    final value = delta / (_state?.context.size?.height ?? 0);
    _state?._dragAnimationController.value += value;
  }

  /// animate panel height to passed value
  Future<void> animateTo(double to) async {
    if (!isAttached) throw UnAttachStateException('you can\'t use controller before _ScrollablePanelState build.');
    await _state?._animateTo(to);
  }

  /// animate panel height to defaultPanelSize
  Future<void> toDefault() async {
    if (!isAttached) throw UnAttachStateException('you can\'t use controller before _ScrollablePanelState build.');
    await _state?._toDefault();
  }

  /// alias toDefault
  Future<void> open() async {
    await toDefault();
  }

  /// animate panel height to maxPanelSize
  Future<void> expand() async {
    if (!isAttached) throw UnAttachStateException('you can\'t use controller before _ScrollablePanelState build.');
    await _state?._expand();
  }

  /// animate panel height to minPanelSize
  Future<void> close() async {
    if (!isAttached) throw UnAttachStateException('you can\'t use controller before _ScrollablePanelState build.');
    await _state?._close();
  }
}

class UnAttachStateException implements Exception {
  final String message;

  UnAttachStateException(this.message);
}

@visibleForTesting
class PanelScrollController extends ScrollController {
  final PanelController controller;

  PanelScrollController({
    double initialScrollOffset = 0.0,
    bool keepScrollOffset = true,
    String? debugLabel,
    required this.controller,
  }) : super(
          keepScrollOffset: keepScrollOffset,
          debugLabel: debugLabel,
          initialScrollOffset: initialScrollOffset,
        );

  @override
  PanelScrollPosition createScrollPosition(ScrollPhysics physics, ScrollContext context, ScrollPosition? oldPosition) {
    return PanelScrollPosition(physics: physics, context: context, oldPosition: oldPosition, controller: controller);
  }
}

@visibleForTesting
class PanelScrollPosition extends ScrollPositionWithSingleContext {
  PanelScrollPosition({
    required ScrollPhysics physics,
    required ScrollContext context,
    double initialPixels = 0.0,
    bool keepScrollOffset = true,
    ScrollPosition? oldPosition,
    String? debugLabel,
    required this.controller,
  }) : super(
          physics: physics,
          context: context,
          initialPixels: initialPixels,
          keepScrollOffset: keepScrollOffset,
          oldPosition: oldPosition,
          debugLabel: debugLabel,
        );

  final PanelController controller;

  bool get listShouldScroll => pixels > 0.0;
  double get controllerValue => controller.animation?.value ?? 0;

  @override
  void applyUserOffset(double delta) {
    if (!listShouldScroll &&
        (!(controllerValue == controller.maxPanelSize || controllerValue == controller.minPanelSize) ||
            (controllerValue < controller.maxPanelSize && delta < 0) ||
            (controllerValue > controller.minPanelSize && delta > 0))) {
      controller._updateExtent(-delta);
    } else {
      super.applyUserOffset(delta);
    }
  }

  @override
  void goBallistic(double velocity) {
    if (!listShouldScroll &&
        (!(controllerValue == controller.maxPanelSize || controllerValue == controller.minPanelSize) ||
            (controllerValue < controller.maxPanelSize && velocity < 0) ||
            (controllerValue > controller.minPanelSize && velocity > 0))) {
      super.goBallistic(0);
      if (controllerValue < controller.maxPanelSize && velocity > 300) {
        controller.animateTo(controller.maxPanelSize);
      } else if (controllerValue > controller.defaultPanelSize) {
        if (velocity < -2000) {
          controller.animateTo(controller.minPanelSize);
        } else if (velocity < -300) {
          controller.animateTo(controller.defaultPanelSize);
        }
      } else if (controllerValue < controller.defaultPanelSize && velocity < -300) {
        controller.animateTo(controller.minPanelSize);
      }
    } else {
      super.goBallistic(velocity);
    }
  }
}


example/ios/Flutter/flutter_export_environment.sh
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export "FLUTTER_ROOT=/Users/ooitakayuki/Develop/flutter"
export "FLUTTER_APPLICATION_PATH=/Users/ooitakayuki/Develop/scrollable_panel/example"
export "FLUTTER_TARGET=/Users/ooitakayuki/Develop/scrollable_panel/example/lib/main.dart"
export "FLUTTER_BUILD_DIR=build"
export "SYMROOT=${SOURCE_ROOT}/../build/ios"
export "OTHER_LDFLAGS=$(inherited) -framework Flutter"
export "FLUTTER_FRAMEWORK_DIR=/Users/ooitakayuki/Develop/flutter/bin/cache/artifacts/engine/ios"
export "FLUTTER_BUILD_NAME=1.0.0"
export "FLUTTER_BUILD_NUMBER=1"
export "DART_DEFINES=flutter.inspector.structuredErrors%3Dtrue"
export "DART_OBFUSCATION=false"
export "TRACK_WIDGET_CREATION=true"
export "TREE_SHAKE_ICONS=false"
export "PACKAGE_CONFIG=.packages"


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:scrollable_panel/scrollable_panel.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({
    Key? key,
    required this.title,
  }) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  PanelController _panelController = PanelController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: PreferredSize(
        preferredSize: const Size.fromHeight(kToolbarHeight),
        child: _AnimatedAppBar(
          panelController: _panelController,
        ),
      ),
      body: Stack(
        children: [
          Positioned.fill(
            child: _FirstView(controller: _panelController),
          ),
          ScrollablePanel(
            defaultPanelState: PanelState.close,
            controller: _panelController,
            onOpen: () => print('onOpen'),
            onClose: () => print('onClose'),
            onExpand: () => print('onExpand'),
            builder: (context, controller) {
              return SingleChildScrollView(
                controller: controller,
                child: _SecondView(),
              );
            },
          ),
        ],
      ),
    );
  }
}

class _AnimatedAppBar extends StatefulWidget {
  const _AnimatedAppBar({
    Key? key,
    required this.panelController,
  }) : super(key: key);

  final PanelController panelController;

  @override
  __AnimatedAppBarState createState() => __AnimatedAppBarState();
}

class __AnimatedAppBarState extends State<_AnimatedAppBar> with SingleTickerProviderStateMixin {
  late final Animation _animation;
  late final AnimationController _animationController;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(vsync: this, duration: const Duration(milliseconds: 300));
    widget.panelController.animation?.addListener(() {
      if (widget.panelController.animation!.value >= 0.8) {
        _animationController.value = (0.2 - (1.0 - widget.panelController.animation!.value)) * 5;
      } else {
        _animationController.value = 0;
      }
    });
    _animation = ColorTween(
      begin: Colors.white,
      end: Colors.red,
    ).animate(_animationController);
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) {
        return Material(
          elevation: 4.0,
          color: _animation.value,
          child: SafeArea(
            bottom: false,
            top: true,
            child: Container(
              height: kToolbarHeight,
              child: const Center(
                child: Text('scrollable panel'),
              ),
            ),
          ),
        );
      },
    );
  }
}

class _FirstView extends StatelessWidget {
  const _FirstView({
    Key? key,
    required this.controller,
  }) : super(key: key);

  final PanelController controller;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        ElevatedButton(
          onPressed: () => controller.open(),
          child: const Text('open panel'),
        ),
        ElevatedButton(
          onPressed: () => controller.close(),
          child: const Text('close panel'),
        ),
        ElevatedButton(
          onPressed: () => controller.expand(),
          child: const Text('expand panel'),
        ),
      ],
    );
  }
}

class _SecondView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    const double circularBoxHeight = 16.0;
    final Size size = MediaQuery.of(context).size;
    return LayoutBuilder(
      builder: (context, constraints) {
        return ConstrainedBox(
          constraints: BoxConstraints(
            minHeight: size.height + kToolbarHeight + 44.0,
          ),
          child: Container(
            decoration: BoxDecoration(
              color: Colors.blue,
              borderRadius: const BorderRadius.only(topLeft: Radius.circular(circularBoxHeight), topRight: Radius.circular(circularBoxHeight)),
              border: Border.all(color: Colors.blue),
            ),
            child: const Center(
              child: Text("second"),
            ),
          ),
        );
      },
    );
  }
}
